[2단계 - 경로 조회 기능] 토리(천은정) 미션 제출합니다.

ezzanzzan opened the pull request:
안녕하세요 찰리 !! 🙇‍♀️

시간이 부족하여 일단 기능 구현에 집중해서 구현하였습니다. 🥲
부족한 테스트 코드 및 예외 사항은 추후 리팩토링 과정에서 추가하도록 하겠습니다 죄송합니다 ㅠㅠ 🥺🥺

리뷰 잘 부탁드립니다 🙏🏻

ezzanzzan committed:
docs: 2단계 요구사항 추가

ezzanzzan committed:
refactor: SectionService 내부 메서드 네이밍 리팩토링

- convertToLineByEntity -> findLine
- convertToSectionByEntity -> findSections

ezzanzzan committed:
feat: NoSuchPath 예외 처리 클래스 생성

ezzanzzan committed:
chore: 데이터베이스 설정

ezzanzzan committed:
feat: 경로 조회 기능 구현

ezzanzzan committed:
feat: 요금 조회 API 및 계산 기능 추가

ezzanzzan committed:
docs: API 명세서 최신화

ezzanzzan committed:
test: 최단 경로 조회 통합 테스트 구현

Gomding commented on src/main/java/subway/controller/PathController.java:
```diff
@@ -0,0 +1,25 @@
+        this.pathService = pathService;
+    }
+
+    @GetMapping("/shortestPath")
```
url 을 카멜케이스로 지정해주신 이유가 궁금해요~

ezzanzzan replied:
API url 네이밍 규칙에 대해서 제대로 파악하지 못해서 카멜케이스로 작성하게 되었어요 🥲
네이밍 규칙에 대해 다시 한 번 정리하며 `paths/shortest-path`로 url을 변경해보았습니다 !!

ezzanzzan replied:
API url 네이밍 규칙에 대해서 제대로 파악하지 못해서 카멜케이스로 작성하게 되었어요 🥲
네이밍 규칙에 대해 다시 한 번 정리하며 `paths/shortest`로 url을 변경해보았습니다 !!

Gomding commented on src/main/java/subway/domain/Path.java:
```diff
@@ -0,0 +1,72 @@
+        return new Path(graph);
+    }
+
+    public List<String> getDijkstraShortestPath(String sourceStation, String targetStation) {
```
`다익스트라 알고리즘`을 사용한다는 부분은 구체적인 사항인것으로 보여요

외부에서는 최소경로를 요청하기만 하면 된다고 생각해요 :)
추상화를 해보면 어떨까요?

Gomding replied:
sourceStation 과 targetStation 이 같다면 어떻게 응답해줘야할까요 🤔

ezzanzzan replied:
경로를 구하는 기능을 추상화하고, 구현체를 통해 다익스트라 알고리즘을 통해 최단경로를 찾는 로직을 구현하였습니다 !!
출발점과 도착점이 같을 경우 예외 처리를 해주는 기능도 함께 추가하였습니다 😊

Gomding commented on src/main/java/subway/domain/Path.java:
```diff
@@ -0,0 +1,72 @@
+            return 0;
+        }
+        return (int) ((Math.ceil((distance - 1) / unit) + 1) * ADD_FARE);
+    }
```
요금 계산에 대한 책임은 분리해보면 어떨까요? 😄

ezzanzzan replied:
경로 탐색 기능을 추상화 하고 요금 계산에 대한 책임을 가지는 Fare 라는 클래스를 생성하여 책임을 분리해 주었습니다 🤓 !!

Gomding commented on src/main/java/subway/domain/Path.java:
```diff
@@ -0,0 +1,72 @@
+            graph.setEdgeWeight(graph.addEdge(
+                    section.getUpStation().getName(),
+                    section.getDownStation().getName()
+            ),section.getDistance());
```
역 이름대신 역(station) 객체를 vertex 로 활용해보면 어떨까요?

ezzanzzan replied:
원래 계획으로는 역 이름 대신 역으로 하려 했는데 시간 관계상 빠르게 구현하려나 보니 이렇게 되었네요 .. 😂

역 객체를 vertext로 활용 할 수 있도록 리팩토링 하였습니다 🤓 !

Gomding commented on src/main/java/subway/domain/Section.java:
```diff
@@ -1,10 +1,11 @@
 import subway.exception.InvalidException;
 
-public class Section {
+public class Section extends DefaultWeightedEdge {
```
도메인에 외부 라이브러리에 대한 상속이 생겼군요? 🤔 
도메인에서는 외부 라이브러리에 대한 의존을 끊어보면 어떨까요?

ezzanzzan replied:
이 부분도 제대로 확인하지 않고 제출했던 부분이네요 죄송합니다 🥺

리팩토링 과정에서 의존을 제거했습니다 !! 🤓

Gomding reviewed (changes requested):
안녕하세요 토리!
2단계 구현하느라 고생하셨어요~
몇 가지 코멘트 남겼으니 확인부탁드려요 🙏 

궁금한 점 있으시면 언제든 DM 남겨주세요!

ezzanzzan committed:
docs: 리팩토링 목록 추가

ezzanzzan committed:
refactor: 최단 경로 조회 API url 변경

ezzanzzan committed:
feat: LineRepository 생성

ezzanzzan committed:
refactor: LineService 리팩토링

- LineRepository 로 메소드 책임 분리
- Line 모델 id 필드 추가

ezzanzzan committed:
style: 공백 제거

ezzanzzan committed:
refactor: SectionService 리팩토링

- SectionRepository 생성
- StationRepository 생성
- SectionService 내 메소드 책임 분리

ezzanzzan committed:
refactor: StationService 리팩토링

- StationService 내 메소드 책임 분리

ezzanzzan committed:
style: 코드 스타일 정렬

ezzanzzan committed:
refactor: Repository 중복 코드 제거

ezzanzzan committed:
fix: Station 생성 시 발생하는 오류 수정

ezzanzzan committed:
refactor: 요금 계산 기능 책임 분리

- Path class -> Fare class 생성

ezzanzzan committed:
refactor: 최단 경로 조회 기능 리팩토링

- 최단 경로 조회 API 응답 형식 변경
- sourceStation과 targetStation이 같은 경우 예외 처리 기능 추가
- graph의 vertex로 Station 객체 사용

ezzanzzan committed:
refactor: Section 도메인 외부 라이브러리 의존 제거

ezzanzzan committed:
refactor: 경로 탐색 기능 추상화

ezzanzzan committed:
test: 최단 경로 조회 기능 테스트 추가

ezzanzzan committed:
feat: 요금 계산 시 예외 처리 기능

- 추가거리가 0 이하일 때 예외 처리 기능 구현

ezzanzzan committed:
test: 요금 계산 기능 테스트 추가

ezzanzzan committed:
style: DisplayName 변경

ezzanzzan committed:
test: ShortestPath 테스트 추가

ezzanzzan committed:
style: DisplayName 변경

ezzanzzan commented:
안녕하세요 찰리 ! 
재리뷰 요청 많이 늦어졌네요.. 죄송합니다 🥺

리팩토링 과정에서 먼저 첫 리뷰 요청을 할 때 시간에 쫓겨 하지 못했던 `요금 계산 기능, 경로 조회 기능에 대한 테스트`를 추가하였습니다 !
그리고 기존 코드에서 Entity -> Domain 변환 과정에서 발생하는 **중복 코드**들을 `Repository`를 사용해서 제거해보았습니다. ˙ᵕ˙ 

미션 동안 좋은 리뷰 해주셔서 정말 감사합니다 🙇‍♀️

ezzanzzan committed:
docs: 문서 업데이트

Gomding commented on src/main/java/subway/domain/ShortestPath.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Override
+    public List<Station> findPath(Station sourceStation, Station targetStation) {
+        DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath = new DijkstraShortestPath<>(graph);
+        GraphPath<Station, DefaultWeightedEdge> path = dijkstraShortestPath.getPath(sourceStation, targetStation);
+        validateSameStation(sourceStation, targetStation);
+        validatePossiblePath(path);
+        return path.getVertexList();
+    }
+
+    @Override
+    public int getDistance(Station sourceStation, Station targetStation) {
+        DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath = new DijkstraShortestPath<>(graph);
+        GraphPath<Station, DefaultWeightedEdge> path = dijkstraShortestPath.getPath(sourceStation, targetStation);
```
경로를 찾을때와 거리를 찾을 때 매번 DijkstraShortestPath 를 초기화하고 사용해야 할까요? 🤔 

DijkstraShortestPath 를 한번 초기화해놓고 재사용할 수 있는 방법은 없을지 고민해보아요 😄

Gomding commented on src/main/java/subway/domain/ShortestPath.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public int getDistance(Station sourceStation, Station targetStation) {
+        DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath = new DijkstraShortestPath<>(graph);
+        GraphPath<Station, DefaultWeightedEdge> path = dijkstraShortestPath.getPath(sourceStation, targetStation);
```
경로를 찾을때와 거리를 찾을 때 매번 DijkstraShortestPath 를 초기화하고 사용해야 할까요? 🤔 

DijkstraShortestPath 를 한번 초기화해놓고 재사용할 수 있는 방법은 없을지 고민해보아요 😄 

+) 
Path 객체는 경로와 거리를 `값`으로 가지고
ShortestPath 는 Sections 정보를 받아서 최소 경로를 Path 로 반환해보는 건 어떻게 생각하시나요?

ezzanzzan replied:
```
private final DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath;
```
로 변경하여 `DijkstraShortestPath`를 한 번 초기화 시켜두고 사용할 수 있도록 리팩토링하였습니다 !

다만 찰리께서 남겨주신 리뷰와는 조금 다른 방향인 듯 해서.. 의도하신 방향이 맞는지 궁금합니다 🥺

Gomding replied:
깔끔하게 리팩토링 해주셨네요 👍 👍 👍 
추가로 해당 부분에 의견 남겨봤습니다 🙌

Gomding commented on src/main/java/subway/domain/ShortestPath.java:
```diff
@@ -0,0 +1,61 @@
+package subway.domain;
+
+import java.util.List;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.shortestpath.DijkstraShortestPath;
+import org.jgrapht.graph.DefaultWeightedEdge;
+import org.jgrapht.graph.WeightedMultigraph;
```
여전히 도메인에 외부 라이브러리에 대한 의존이 존재하고 있어요..! 😃 

`jgrapht` 라이브러리를 사용해 다익스트라 알고리즘을 사용한 최소경로를 구하기 는 어느 layer 에 속한다고 생각하시나요? 🤔 
저는 `jdbc` 를 사용해 저장소에 접근하기 와 비슷한 결이라고 생각해요!

ezzanzzan replied:
앗 제가 코멘트의 의도를 제대로 파악하지 못했었군요 🥺

> 저는 jdbc 를 사용해 저장소에 접근하기 와 비슷한 결이라고 생각해요!

`jgrapht` 외부 라이브러리를 쓰는건 `jdbc` 쓰는거랑 마찬가지니, 이 외부 라이브러리가 사용되는 부분에 대해 패키지를 따로 분리해보는게 어떨까요? 라는 의도로 이해했는데 제가 제대로 이해한게 맞을까요 ?! 🥺
 
해당 클래스를 Repository 패키지로 이동하고 Path 객체를 새로 생성하여 Path 객체를 반환할 수 있도록 리팩토링 하였습니다.
레이어드 아키텍처에 대한 이해도가 부족하다고 생각해서 계속 학습중인데 이런 방향성을 의도하신게 맞을까요 .. ! 🥺

Gomding replied:
정확합니다!! 😄 

나중에 경로 조회에 사용하는 라이브러리가 교체되더라도 도메인은 영향을 받지 않는것이 좋다고 생각해요~
경로 조회를 위해 어떤 라이브러리를 사용할지 어떤 알고리즘을 사용할지는 도메인에서 궁금하지 않아도 된다고 생각합니다 ㅎㅎ
경로 값만 얻을 수 있으면 되니까요..!

추가로 repository 의 개념이 맞는지는 의문이 들었어요!
토리가 repository를 어떤 layer에 속한다고 생각하는지 궁금하군요 😄 
조금 더 가자면 infrastructure 라는 레이어도 있으니 학습해보면 좋겠어요~

Gomding commented on src/main/java/subway/exception/NoSuchPath.java:
```diff
@@ -0,0 +1,9 @@
+package subway.exception;
```
`GlobalControllerAdvice` 클래스 파일의 코드 위치에 코멘트를 남길 수 없어서 여기에 남깁니다 🙏 

예외를 핸들링할 때 Logger 를 사용하여 로그를 적절하게 남겨보는것은 어떠신가요? 😄

ezzanzzan replied:
예외 발생 시 로그로 확인할 수 있도록 Logger를 추가해주었습니다 !!

Gomding commented on src/main/java/subway/exception/ErrorCode.java:
```diff
@@ -4,6 +4,7 @@ public enum ErrorCode {
     DUPLICATE_NAME("중복된 이름입니다."),
     DUPLICATE_SECTION("이미 연결되어 있는 구간입니다."),
     NO_SUCH_STATION("존재하지 않는 역입니다."),
+    NO_SUCH_PATH("존재하지 않는 경로입니다."),
```
존재하지 않은 경로라고 반환해주는 것은 좋지만, 해당 정보만으로 예외가 발생했을 때 상황 추적이 가능할까요? 🤔 
출발역과 도착역이 무었이었길래 존재하지 않는 경로예외가 발생했을까? 라는 의문이 생기지 않을까요!

경로가 진짜 잘못됐는지 빠르게 알 수 있도록 조회를 시도한 출발역과 도착역도 알려주는 것은 어떨까요?

ezzanzzan replied:
존재하지 않는 역과 경로로 인한 예외가 발생했을 때 어떤 상황인지 조금 더 빠르게 판단하고 추적하기 위해서
어떤 역인지, 어떤 경로인지 추가로 전달하여 좀 더 자세한 예외 메시지를 확인할 수 있도록 하였습니다 !! 🫡

Gomding replied:
👍 👍 👍 

예외 메시지에 대해 참고할 수 있는 아티클 공유드립니다!! 😄 
https://hyeon9mak.github.io/writing-helpful-error-messages/

Gomding commented on src/main/java/subway/repository/LineRepository.java:
```diff
@@ -0,0 +1,46 @@
+
+@Repository
+public class LineRepository {
+    private final LineDao lineDao;
+    private final SectionRepository sectionRepository;
```
Repository 와 Dao 의 차이는 무엇일까요 🤔

Gomding commented on src/main/java/subway/service/PathService.java:
```diff
@@ -0,0 +1,45 @@
+        int shortestPathDistance = path.getDistance(sourceStation, targetStation);
+        int shortestPathFare = fare.calculateFare(shortestPathDistance);
+
+        List<StationResponse> stationResponses = shortestPath.stream()
```
StationResponse 의 경우 여러 API 에서 사용되는것으로 보여요!

![image](https://github.com/woowacourse/jwp-subway-path/assets/57378410/f91689c5-0b1a-4a82-b0a2-92788a58b40b)

Response 값은 굉장히 변하기 쉬운 부분인데요
표현하는 값이 같아서 동일한 Response 를 쓰다보면
한 부분에서 응답값에 변화가 있어야할 때 다른 곳도 영향을 받을 수 있어요.

물론 동일한 값을 가지는 Response 들을 만드는것이 비효율적으로 보일 수 있지만
미래를 생각하면 분리하는 편이 좋다고 생각해요!
각각의 응답은 현재는 같은 길을 걷고있지만 언제든지 갈라질 수 있어요 ㅎㅎ

토리는 어떻게 생각하시나요?

ezzanzzan replied:
일단 기존에는 생각해보지 못했던 부분이었습니다 !
표현하는 값이 같아 이미 생성되어 있는 Response를 큰 고민없이 재사용하고 있었네요...🧐

저도 찰리의 의견처럼 현재 상황에서는 각각의 Response를 만드는 것이 비효율적일 수 있다고 생각하지만, 변화를 생각한다면 각각의 응답을 나타내는 Response가 분리되어 있는 것이 좋다고 생각했습니다 !! 
Response 자체를 나누는 것이 큰 비용이 발생하는 일도 아니라고 생각이 들었구요 ˙ᵕ˙ 
오히려 분리해두지 않았다 추후 변경이 발생했을 때 다른 곳에 영향이 생기는 것보다 훨씬 적은 비용이라고 생각했습니다. 🥺

`LineStationResponse`, `PathStationResponse` 로 분리하는 방향으로 리팩토링을 진행하였습니다. 🫡

Gomding commented on src/main/java/subway/domain/Line.java:
```diff
@@ -4,19 +4,21 @@
 import java.util.List;
 
 public class Line {
+    private final Long id;
```
Line 객체에 id 가 갑자기 생겨났군요!
어떤 과정에서 다시 id를 가져야겠다 생각하셨을까요! 😄

ezzanzzan replied:
리팩토링 전에는 `Repository`가 존재하지 않고 전체 라인을 조회할 때 `Dao`에서 가져온 `lineEntity`를 통해 바로 `lineResponse`를 생성해서 응답을 보내주는 형식이었습니다 ! 
그래서 `Line` 객체에 `id`가 없어도 다른 조회과정이 필요하지 않았습니다.

리팩토링 과정에서 `Repostiory`를 추가하게 되면서 `Repository`에서 `Line` 객체를 생성하고 `Service`로 `Line` 객체를 보내준 뒤 해당 객체를 통해 `Response`로 변환하는 과정에서 `id` 조회 과정이 필요해졌고 이런 이유로 `Line` 객체에도 `id`를 추가하게 되었습니다 !! 🫡

Gomding commented on src/main/java/subway/controller/PathController.java:
```diff
@@ -0,0 +1,25 @@
+        this.pathService = pathService;
+    }
+
+    @GetMapping("/shortest")
```
url path 를 적절하게 변경 👍

Gomding commented on src/main/java/subway/domain/Fare.java:
```diff
@@ -0,0 +1,42 @@
+        this(DEFAULT_FARE);
+    }
+
+    private void validateDistance(int distance) {
+        if (distance <= 0) {
+            throw new InvalidException(ErrorCode.INVALID_NOT_POSITIVE_DISTANCE);
+        }
+    }
```
메서드 위치가 호출하는 곳보다 위에 있군요 🦅

ezzanzzan replied:
메소드를 정렬하는 과정에서 계산 로직은 계산 로직대로, 검증 로직은 검증 로직대로 모아두고 싶어져서 메소드 위치를 이동시키다보니 호출하는 곳보다 위로 올라가버렸네요... 🥲 
다시 생각해보니 기존 다른 클래스들에서 하던 메소드 순서대로 정렬하는 것이 통일성도 있고 코드를 읽어나가는데도 불편함이 없을 것 같아 다시 메소드에서 호출되는 순서대로 정렬을 진행하였습니다 !! 🫡

Gomding commented on src/main/java/subway/domain/Fare.java:
```diff
@@ -0,0 +1,42 @@
+        return fare;
+    }
+
+    private int calculateOverFare(int distance, int unit) {
+        if (distance < 0) {
+            return 0;
+        }
```
`validateDistance` 에서 distance 가 음수인것을 검사하여 예외를 던지는데
여기서는 음수일 때 0을 반환하는 군요 🤔 
이동 거리가 음수인 것이 `정상적인 흐름 VS 예외 상황` 어느 쪽으로 생각해야할지 혼란이 생길것같아요 😄

ezzanzzan replied:
처음 들어왔을 때 거리에 대한 값을 검사 한 뒤, 계산 로직 상에서 `거리가 0보다 작으면 더해질 요금 값이 없다` 라는 의미로 구현을 하고 검증 로직과 혼돈이 일어날 수도 있다는 생각은 하지 못했었습니다 🥺
찰리의 코멘트를 보고 `정상적인 흐름 VS 예외 상황` 사이에서 충분히 혼돈이 일어날 수 있다고 생각해서 조건문을 추가하고 해당 부분 로직을 제거하는 방향으로 리팩토링을 진행하였습니다 !! 🫡

Gomding reviewed (changes requested):
안녕하세요 토리!
2단계 피드백 반영하느라 고생하셨습니다 👏 
추가로 몇가지 코멘트 남겼으니 확인 부탁드려요!!

궁금한 점 있으면 언제든 DM 남겨주세요 🙏

ezzanzzan committed:
style: 패키지 수정

ezzanzzan committed:
docs: 리팩토링 목록 업데이트

ezzanzzan committed:
refactor: 요금 계산 로직 리팩토링

ezzanzzan committed:
refactor: 응답에 따라 Response 분리

ezzanzzan committed:
refactor: 오류 메시지 상세화

ezzanzzan committed:
refactor: 외부 라이브러리 의존 제거

ezzanzzan committed:
feat: 예외 핸들링 시 Logger 추가

ezzanzzan committed:
refactor: 클래스 네이밍 변경

Gomding commented on src/main/java/subway/domain/Fare.java:
```diff
@@ -19,24 +20,28 @@ public Fare() {
+    private void validateDistance(int distance) {
+        if (distance <= 0) {
+            throw new InvalidException(ErrorCode.INVALID_NOT_POSITIVE_DISTANCE);
         }
```
훌륭하게 리팩토링 해주셨네요~!! 👍

Gomding commented on src/main/java/subway/repository/StationRepository.java:
```diff
@@ -42,7 +42,7 @@ public Station findByName(String name) {
             return new Station(stationEntity.getId(), stationEntity.getName());
         } catch (Exception e) {
             System.out.println(e.getMessage());
```
디버깅용 콘솔 출력이 일까요? 👀

Gomding commented on src/main/java/subway/domain/PathFinder.java:
```diff
@@ -0,0 +1,14 @@
+import subway.exception.ErrorCode;
+import subway.exception.InvalidException;
+
+public interface PathFinder {
```
ShortestPathFinder 가 아닌 PathFinder 라는 이름으로 추상화 해주신것은 
다른 종류의 경로 조회가 나올것을 생각해주신 것일까요? 😄

Gomding commented on src/main/java/subway/repository/ShortestPathFinder.java:
```diff
@@ -0,0 +1,44 @@
+import subway.exception.ErrorCode;
+import subway.exception.NoSuchPath;
+
+public class ShortestPathFinder implements PathFinder {
+    private final DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath;
```
빈으로 관리해보는것은 어떨까요? 😄
graph 를 한 번 초기화 해놓고 재사용할 수 있도록요

지하철 노선도 라는 도메인 특성상 노선 추가, 수정, 삭제 보다는
노선 경로 조회 요청이 대부분일 거라 생각해요
그럼 조회할 때 마다 graph 를 새롭게 만드는 과정은 비효율적이지 않을까요~

그럼 ShortestPathFinder 를 빈으로 등록하면서 graph 를 초기화 시켜놓을 수 있겠네요~
(마치 캐싱처럼)

여기서 추가로 고민이 나올 수 있어요
1. graph 를 캐싱해놓은건 좋은데 노선 추가, 수정, 삭제가 발생하면 graph 를 어떻게 수정해줘야할까?
2. 1번은 잘 만드려면 또 고민해야할 것이 멀티 스레드 환경에서 graph 에 수정이 일어났을 때는 어떻게 해야할까~

정도의 고민이 있곘네요 😄

Gomding commented on src/main/java/subway/exception/GlobalControllerAdvice.java:
```diff
@@ -8,19 +10,24 @@
     @ExceptionHandler(MethodArgumentNotValidException.class)
     public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
         String errorMessage = e.getBindingResult().getAllErrors().get(0).getDefaultMessage();
+        LOGGER.info(errorMessage);
```
에러 핸들링 시 Logging 레벨은 어떻게 관리해주는 것이 좋을까요~
물론 어떤 레벨을 사용하는것에 정답은 없어서
토리만의 기준을 세워보시면 좋겠어요
(팀에 가면 팀의 컨벤션도 있을겁니다 ㅎㅎ)

[로깅 관련 테코톡](https://www.youtube.com/watch?v=1MD5xbwznlI&t=13s)

Gomding commented on src/main/java/subway/domain/Fare.java:
```diff
@@ -0,0 +1,38 @@
+import subway.exception.ErrorCode;
+import subway.exception.InvalidException;
+
+public class Fare {
```
Fare 라는 이름은 VO 를 연상시킬 수 있다고 생각해요~

`요금 계산기` 라고 인지할 수 있도록 이름을 부여해보는 것은 어떨까요? 😄

Gomding commented on src/main/java/subway/domain/Fare.java:
```diff
@@ -0,0 +1,38 @@
+        validateDistance(distance);
+        int fare = DEFAULT_FARE;
+
+        if (distance > 50) {
+            fare = DEFAULT_FARE_OVER_50KM;
+            fare += calculateOverFare(distance - 50, UNIT_OVER_50KM);
+        }
+
+        if (distance >= 10 && distance <= 50) {
+            fare += calculateOverFare(distance - 10, UNIT_OVER_10KM);
+        }
```
해당 매직 넘버들은 어떤 의미를 가지고 있을까요~

Gomding commented on src/test/java/subway/integration/ShortestPathIntegrationTest.java:
```diff
@@ -0,0 +1,140 @@
+        assertThat(response.as(PathResponse.class).getFare()).isEqualTo(1250);
+    }
+
+    @DisplayName("출발역과 도착역이 같을 때 예외가 발생한다.")
+    @Test
```
꼼꼼한 예외처리 테스트 👍

Gomding reviewed (approved):
안녕하세요 토리!
고민하며 피드백 반영해주셔서 감사해요 👍 
지하철 미션은 이만 merge 해도 되겠어요~
추가로 남긴 코멘트에서 궁금한 점 있으시면 DM 보내주세요 😄 
고생많으셨습니다 🥳
